File: buck/dev/clippy.py
# clippy over the workspace through buck2, which is what //:verify's lint pass
# runs.
#
# buck/dev/workspace.bxl builds every first-party Rust target's [clippy.json] -
# rustc's JSON diagnostics - in every configuration the workspace builds it in,
# on BuildBuddy's workers. This reads them. A file compiled in more than one
# configuration or target is diagnosed more than once, so each finding is
# reported once. The toolchain denies warnings, so any finding fails the run.
#
# With --fix it first applies the suggestions clippy marks machine-applicable,
# which is what `cargo clippy --fix` applies: every edit a diagnostic suggests
# goes in together or not at all, one that overlaps an edit already taken is
# left for the next run, and then clippy runs again over the result, so what is
# reported is what the fixes did not reach.

import json
import os
import subprocess
import sys


def diagnostics(buck):
    listing = subprocess.run(
        [buck, "bxl", "//buck/dev/workspace.bxl:subtarget", "--", "--subtarget", "clippy.json"],
        check=True,
        stdout=subprocess.PIPE,
        text=True,
    ).stdout
    found = []
    for line in listing.splitlines():
        if "\t" not in line:
            continue
        _, path = line.split("\t", 1)
        with open(path) as file:
            for entry in file:
                entry = entry.strip()
                if entry:
                    found.append(json.loads(entry))
    return found


def first_party(diagnostic):
    return any(span["file_name"].startswith("crates/") for span in diagnostic.get("spans", []))


def suggestions(diagnostic):
    edits = []
    pending = [diagnostic]
    while pending:
        current = pending.pop()
        for span in current.get("spans", []):
            if span.get("suggested_replacement") is None:
                continue
            if span.get("suggestion_applicability") != "MachineApplicable":
                continue
            edits.append((span["file_name"], span["byte_start"], span["byte_end"], span["suggested_replacement"]))
        pending.extend(current.get("children", []))
    return edits


def fix(found):
    chosen = {}
    applied = 0
    seen = set()
    for diagnostic in found:
        edits = sorted(set(suggestions(diagnostic)))
        if not edits or not all(edit[0].startswith("crates/") for edit in edits):
            continue
        key = tuple(edits)
        if key in seen:
            continue
        seen.add(key)
        overlaps = any(
            start < other_end and other_start < end
            for path, start, end, _ in edits
            for other_start, other_end, _ in chosen.get(path, [])
        )
        if overlaps:
            continue
        for path, start, end, replacement in edits:
            chosen.setdefault(path, []).append((start, end, replacement))
        applied += 1
    for path, edits in chosen.items():
        with open(path, "rb") as file:
            source = file.read()
        for start, end, replacement in sorted(edits, reverse=True):
            source = source[:start] + replacement.encode() + source[end:]
        with open(path, "wb") as file:
            file.write(source)
    return applied


def report(found):
    rendered = []
    for diagnostic in found:
        if diagnostic.get("level") not in ("error", "warning") or not first_party(diagnostic):
            continue
        text = diagnostic.get("rendered") or diagnostic["message"]
        if text not in rendered:
            rendered.append(text)
    for text in rendered:
        sys.stdout.write(text)
    return len(rendered)


def run(buck, fixing):
    found = diagnostics(buck)
    if fixing and found:
        applied = fix(found)
        if applied:
            print("Applied the fixes of {} clippy findings.".format(applied))
            found = diagnostics(buck)
    count = report(found)
    if count:
        print("clippy: {} findings.".format(count))
    return count == 0

File: crates/buck-tools/src/clippy.rs
use std::collections::{BTreeMap, BTreeSet};
use std::process::{Command, ExitCode};

use serde_json::Value;

type Edit = (String, u64, u64, String);

pub fn run(arguments: &[String]) -> Result<ExitCode, String> {
    let (buck, fixing) = match arguments {
        [buck] => (buck, false),
        [buck, flag] if flag == "--fix" => (buck, true),
        _ => return Err("usage: clippy BUCK [--fix]".into()),
    };
    let mut found = diagnostics(buck)?;
    if fixing && !found.is_empty() {
        let applied = fix(&found)?;
        if applied > 0 {
            println!("Applied the fixes of {applied} clippy findings.");
            found = diagnostics(buck)?;
        }
    }
    let count = report(&found);
    if count > 0 {
        println!("clippy: {count} findings.");
        return Ok(ExitCode::FAILURE);
    }
    Ok(ExitCode::SUCCESS)
}

fn diagnostics(buck: &str) -> Result<Vec<Value>, String> {
    let listing = crate::output(Command::new(buck).args([
        "bxl",
        "//buck/dev/workspace.bxl:subtarget",
        "--",
        "--subtarget",
        "clippy.json",
    ]))?;
    let mut found = Vec::new();
    for line in listing.lines() {
        let Some((_, path)) = line.split_once('\t') else {
            continue;
        };
        for entry in crate::read(path)?
            .lines()
            .map(str::trim)
            .filter(|entry| !entry.is_empty())
        {
            found.push(serde_json::from_str(entry).map_err(|error| format!("{path}: {error}"))?);
        }
    }
    Ok(found)
}

fn spans(diagnostic: &Value) -> impl Iterator<Item = &Value> {
    diagnostic["spans"].as_array().into_iter().flatten()
}

fn file_name(span: &Value) -> &str {
    span["file_name"].as_str().unwrap_or_default()
}

fn first_party(diagnostic: &Value) -> bool {
    spans(diagnostic).any(|span| file_name(span).starts_with("crates/"))
}

fn suggestions(diagnostic: &Value) -> BTreeSet<Edit> {
    let mut edits = BTreeSet::new();
    let mut pending = vec![diagnostic];
    while let Some(current) = pending.pop() {
        for span in spans(current) {
            let Some(replacement) = span["suggested_replacement"].as_str() else {
                continue;
            };
            if span["suggestion_applicability"] != "MachineApplicable" {
                continue;
            }
            edits.insert((
                file_name(span).to_owned(),
                span["byte_start"].as_u64().unwrap_or_default(),
                span["byte_end"].as_u64().unwrap_or_default(),
                replacement.to_owned(),
            ));
        }
        pending.extend(current["children"].as_array().into_iter().flatten());
    }
    edits
}

fn fix(found: &[Value]) -> Result<usize, String> {
    let mut chosen: BTreeMap<String, Vec<(u64, u64, String)>> = BTreeMap::new();
    let mut seen = BTreeSet::new();
    let mut applied = 0;
    for diagnostic in found {
        let edits = suggestions(diagnostic);
        if edits.is_empty()
            || !edits.iter().all(|edit| edit.0.starts_with("crates/"))
            || !seen.insert(edits.clone())
        {
            continue;
        }
        let overlaps = edits.iter().any(|(path, start, end, _)| {
            chosen
                .get(path)
                .into_iter()
                .flatten()
                .any(|(other_start, other_end, _)| start < other_end && other_start < end)
        });
        if overlaps {
            continue;
        }
        for (path, start, end, replacement) in edits {
            chosen
                .entry(path)
                .or_default()
                .push((start, end, replacement));
        }
        applied += 1;
    }
    for (path, mut edits) in chosen {
        let mut source = std::fs::read(&path).map_err(|error| format!("{path}: {error}"))?;
        edits.sort();
        for (start, end, replacement) in edits.into_iter().rev() {
            source.splice(start as usize..end as usize, replacement.into_bytes());
        }
        std::fs::write(&path, source).map_err(|error| format!("{path}: {error}"))?;
    }
    Ok(applied)
}

fn report(found: &[Value]) -> usize {
    let mut rendered: Vec<&str> = Vec::new();
    for diagnostic in found {
        let level = diagnostic["level"].as_str().unwrap_or_default();
        if !matches!(level, "error" | "warning") || !first_party(diagnostic) {
            continue;
        }
        let text = diagnostic["rendered"]
            .as_str()
            .or_else(|| diagnostic["message"].as_str())
            .unwrap_or_default();
        if !rendered.contains(&text) {
            rendered.push(text);
        }
    }
    for text in &rendered {
        print!("{text}");
    }
    rendered.len()
}
